### ISSUE TITLE
initial hs_state.mseq_s value

### ISSUE THREAD

**Issue by boaks** [eclipse/tinydtls#84]
According [RFC6347,  4.2.1.  Denial-of-Service Countermeasures, page 17](https://datatracker.ietf.org/doc/html/rfc6347#page-17)

> When the second ClientHello is received, the server can verify that
> the Cookie is valid and that the client can receive packets at the
> given IP address.  In order to avoid sequence number duplication in
> case of multiple cookie exchanges, the server MUST use the record
> sequence number in the ClientHello as the record sequence number in
> its initial ServerHello.  Subsequent ServerHellos will only be sent
> after the server has created state and MUST increment normally.

The implementation in `dtls.c, handle_handshake_msg, line 3631` uses

```
peer->handshake_params->hs_state.mseq_r = dtls_uint16_to_int(hs_header->message_seq);
peer->handshake_params->hs_state.mseq_s = 1;
```

I think, the initial value for `hs_state.mseq_s` value should be also `dtls_uint16_to_int(hs_header->message_seq)`. A test, with an expired cookie and second hello_verify should demonstrate, if my assumption is right.

**Comment by boaks** [comment]
A few lines above the initialization of `hs_state.mseq_s`:

```
security->rseq = 1;
```

FMPOV, the same applies to the record sequence number. The initial value is that of the correlated client hello. Otherwise, client_hello retransmissions/cookie retries will cause the server side to fail the handshake, if the other peer checks the record sequence number window.

**Comment by boaks** [comment]
And that (record sequence number and message sequence number from CLIENT_HELLO) must also be applied for the HELLO_VERIFY_REQUEST.  Otherwise a handshake on the cookie "exchange" will fail. That may be not too often, but in order to process that, the sequence numbers of the CLIENT_HELLO has to be used for the HELLO_VERIFY_REQUEST and SERVER_HELLO.

**Comment by boaks** [comment]
Considering also issues #45 and #71, my intention would be split the "stateless processing" into separate functions would provide a solution.

Checking for CLIENT_HELLOs of epoch 0, "fork" that processing very early, and use the record and message seqn then for both the HELLO_VERIFY_REQUEST or for the SERVER_HELLO and to initialize a new peer, would implement it proper. The current record sequence number reset will then be obsolete.

**Comment by boaks** [comment]
Fixed with PR #89
